### ISSUE TITLE
Multi File Upload 

### ISSUE THREAD

**Issue by Jonathon-Marc** [alphagov/govuk-design-system-backlog#295]
## What
> I am looking to propose a Multi File upload pattern that currently does not exist in GDS. Currently, there are variations from HMRC and the MoJ which is most closely related to our use case for UKHSA. I have contacted govuk-design-system-support@digital.cabinet-office.gov.uk. who asked me to start a new issues. 

## Why
> Explain why you think this should be added to the GOV.UK Design System.

1. The pattern would allow users to upload multiple files at once, review the uploaded files, and remove unwanted files, in it's MVP format with possible other features being explored. 
2. There is currently no other pattern in GDS that meets a users need to upload a bulk amount of files at once. 
3. There is also a current issue on GitHub 'upload file component has no visual target area when dragging and dropping a file'. 
4. We have use cases where users need to upload multiple files at once, over 100 files in one submission. After our research, the NPIS team and the user-centred design team from UKHSA have concluded that no component in GDS meets the users needs.

> - What evidence do you have that it's needed by multiple services across the government? 
Research shows multiple file upload patterns being used by HMRC and MoJ. There are further use cases within UKHSA. 

> - What evidence do you have that it meets the needs of the users of those services?
The file upload component proposal build on what is being used by the MoJ and addresses accessibility concerns from the MoJ patterns. For examples, the use of green and red colours to distinguish if a file is ready uploaded and ticks and cross icons that could present problems with screen readers. Futher research has shows users want to know more details on upload space available and the number of files uploaded giving more clarity to the user. 

> - Have you checked that it doesn't already exist in the GOV.UK Design System? 
Yes

## Anything else
> Include links to any examples, research or code to support your proposal, if available.

<img width="1440" alt="03_Upload files Flow 50" src="https://github.com/alphagov/govuk-design-system-backlog/assets/170522445/368fd155-5b0a-4023-a4d0-d6c9dedc3bbd">
<img width="1440" alt="03_Upload files Flow 66" src="https://github.com/alphagov/govuk-design-system-backlog/assets/170522445/261d3029-61ec-4f01-8060-cb8ced26edd2">
<img width="1440" alt="03_Upload files Flow 116" src="https://github.com/alphagov/govuk-design-system-backlog/assets/170522445/a89eebcd-458b-412f-b22a-00361813a373">

**Comment by Jonathon-Marc** [comment]
<img width="1440" alt="03_Upload files Flow_Label Variants" src="https://github.com/alphagov/govuk-design-system-backlog/assets/170522445/48526da7-920b-4850-9948-a0d22c90e2ab">
[File upload research insights.pptx](https://github.com/user-attachments/files/15567966/File.upload.research.insights.pptx)


Updating this with round 1 UR insights and a recent iteration of the copy to show the status of the file. 
_______________________


Had a helpful conversation with a colleague from Of Qual who is in the alpha phase of a multi-file upload with similar use cases who is using a slightly different approch 
<img width="1145" alt="image (5) 1" src="https://github.com/alphagov/govuk-design-system-backlog/assets/170522445/a91ff478-5e3f-4134-9f03-6478b7dfbe02">

**Comment by Jonathon-Marc** [comment]
[Interaction Design Onboarding.pdf](https://github.com/user-attachments/files/17114316/Interaction.Design.Onboarding.pdf)
<img width="1440" alt="04 1_File Uploader GB Only" src="https://github.com/user-attachments/assets/4a8867f7-1451-431e-90bb-050105bdcb30">
<img width="1440" alt="04 1_File Uploader GB Only-1" src="https://github.com/user-attachments/assets/a4eba226-b7a3-42c8-ad9a-af923ab159bd">
<img width="1440" alt="04_File Uploader GB Only" src="https://github.com/user-attachments/assets/174959dd-a515-4787-bb6e-6937ccc87de6">
<img width="1440" alt="Upload files_Maximum file upload_Error (user continues to upload files)" src="https://github.com/user-attachments/assets/35f2b0be-ca50-4a47-871f-77dcf5a7ed20">
<img width="1440" alt="04_Submission complete" src="https://github.com/user-attachments/assets/25af51be-c229-45b7-b0f1-e5c4ecee2f54">
<img width="1440" alt="Incorrect File GB" src="https://github.com/user-attachments/assets/caa3386b-a5e6-4932-b4e5-483a6e6dbe35">

Following our second round of UR we simplified the content on the page and the hierarchy of the file upload page. 

Consideration for our future backlog is a toast notification when a file has been deleted.

**Comment by diroussel** [comment]
Can I make a few suggestions, comments and raise some discussion points, based on working on a very similar design for a UK public corporation (So government owned, but not part of the government).

1) Drop zone. I like the drag and drop zone with the dashed edge, because it's some how now part of the vernacular of drag and drop components.  But it doesn't look right with square corners.  Just a small rounding of the corners makes the dashes work better, and makes it even clearer that it's a drag and drop target.

2) Drag and drop. If we are going to have this part of the design system, then there should be guidance about how the drag and drop behaves in a dynamic way, not just how it looks.  In particular I had issues with the default behaviour of the browser.  If you drag and drop an image or PDF and miss the drop zone by a small amount, then the browser takes it as an instruction to open the image or PDF.  Either in a new tab, or the current tab.  This is counter to what the user is trying to acheive.  My solution was on any page that includes a file upload component, disable the default drag and drop behaviour, and also highlight the file upload component when the drag hover is above the component to reassure that they are dragging to the right place.  You can see this on other drag and drop implementations out there.

3) File types. When using the input element with type=file, the developer can restrict which file extensions can be picked, for instance only .jpg, .png and .docx.  The user can see other files in the file picker, but they are greyed out and can't be selected.  This is a clear visual guidance to the user.  But when file are drag and dropped this filtering doesn't occur.  So if 5 files are dragged on, and two are not-allowed file types, then what should we do?  Perhaps silently ignore them? Or show a toast notification?  Or add them to the file upload list, but with the standard error bars and message?

4) File sizes.  The examples on the issue page are not precise about file sizes should be displayed.  I think we should use KB, MB, GB using SI units (megabyes, not mebibytes).  Also do we really need a column for the file size?  The file name is more important, so you can just display:  `John's tax records 2022-2023 (1).pdf (2MB)`

5) File naming, should we have guidance on which characters are acceptable for file names?  I would suggest to make sure we can include french, german, polish, cyrilic characters along with the regular english-latin alphabet.

6) Upload progess.  In my system we are dealing with a complex business domain, so some supporting documents might be quite big, particularly if they are scanned images in a PDF.  Behind the scenes we break the file into 1MB chunks, and do a series of chunked uploads.  This can take time if the user is on a slower connection, for instance in a rural setting.  So we must be sure to give visual feedback on the success of the upload.  I have implemented this as a subltle progress bar in the background of the status Tag component.  So the status text changes from PREPARING, STARTING, UPLOADING, PROCESSING to UPLOADED, inside the Tag component, with a suitable colour.  Meanwhile the background moves like a progress bar as each step of the upload completes.

7) File list order.  It's interesting to see that you have the select/dragdrop area at the top of the list. Whereas we have it at the bottom. When we add new files we add them to the bottom of the list. In your prototype do you drag them to the top, but they appear at the bottom?  Or do you also add new items to the top?  I don't think either one is more right. But I currently like the add-to-bottom approach.

8) Malware scanning.  The solution I'm working on incorporates malware scanning.  Unfortunatly this is an async backend process.  So the frontend upload the file to 'PROCESSING' state, then later when we know the malware scan succeed, we update the status to 'UPLOADED'.  If the malware scan detects problems, then we update the status to 'FAILED', and the user must remove the file before completing the section.  If the malware scan hasn't completed yet, the user can leave the page, but can't complete the section.

9) Parallel upload.  We currently upload up to 2 files in parallel.  And in each file up to two chunks at a time.  If a user selects 10 files, then visually they all apear on the page, but with status 'PREPARING' while waiting their turn to upload.

That's all for now.  It's a suprisingly complex topic.

**Comment by Jonathon-Marc** [comment]
Here is our Beta iteration enhanced for accessibility: 

![Image](https://github.com/user-attachments/assets/334d1aec-c75a-4dfe-9d25-870d24c3da66)
![Image](https://github.com/user-attachments/assets/e11ffba3-6680-4956-b9cc-794f9fdd2dbc)
![Image](https://github.com/user-attachments/assets/544ea50d-e71c-405e-b7ab-20e61cc99c93)
![Image](https://github.com/user-attachments/assets/f2f13ac8-47fc-4cd5-a6f5-cf0e608f45e8)

**Comment by Jonathon-Marc** [comment]
![Image](https://github.com/user-attachments/assets/9c3fe562-1bfd-4877-af7c-da768cfd81c8)
![Image](https://github.com/user-attachments/assets/bbeddb18-24b1-49e7-9192-c1b23c2d07eb)
![Image](https://github.com/user-attachments/assets/a71b2108-3bc9-4997-85a3-c4c1502b8d2f)

We also did some future thinking for notifications.

**Comment by Jonathon-Marc** [comment]
https://github.com/user-attachments/assets/e1adc19e-545d-43ce-8e3a-085093969768

**Comment by elken** [comment]
This is really great.

We in MoJ have also done a similar component to enhance multi file uploads (more featureful than https://design-patterns.service.justice.gov.uk/components/multi-file-upload/, matching what @Jonathon-Marc has demoed)

Curious what the priority of adding this in is? Our design system team have recently made some large-scale changes to all the components we use and that has caused us to have to maintain our own version of the multi-file upload; would be ideal if we could use this one instead! 

Thanks all for the great work, demo of ours below for posterity.

https://github.com/user-attachments/assets/11e7aba5-f4ca-4ef9-a891-366f2869e55c

**Comment by CharlotteDowns** [comment]
Hi folks, for the next few weeks we're reviewing [the improved version of the File upload component in the GOV.UK Design System](https://design-system.service.gov.uk/components/file-upload/#using-the-improved-file-upload-component). Get in touch if you're using it, and if you want to suggest any improvements or findings from user research :pray:.

I think there's some limitations to add multi file upload to the component itself but [we may be able to investigate the pattern approach](https://github.com/alphagov/govuk-design-system/issues/5232).

**Comment by elken** [comment]
Thanks @CharlotteDowns! I've flagged this with the team who implemented the changes in my comment so hopefully they reach out for their input

**Comment by akinshaywai** [comment]
Hi @CharlotteDowns, thanks for actively seeking feedback on this.

I'm a designer also working with DWP Universal Credit. Claimants (benefit recipients) regularly struggle with uploading documents, and here are some patterns and pain points I can share from that context:

**1. Error recovery is the real design challenge**
The demos shared above focus heavily on the "happy path." But in practice, the most critical UX moments are:

- What happens when a user uploads 10 files and 2 fail? Is the error state clear enough per file without creating visual noise?
- Can users retry individual failed uploads without re-uploading everything?
- How do we handle file type mismatches in a drag-and-drop context (as @diroussel raised)?

I'd advocate for inline, per-file error states with clear recovery actions, not toast notifications which are easy to miss, especially for screen reader users.

2. Progressive disclosure for complexity
Multi-file upload introduces significant cognitive load for users who may already be stressed (benefits applications, compliance submissions, etc.). From my experience, a progressive approach works well: start with the simple single-file upload, and only reveal the multi-file interface when the user actively needs it. This avoids overwhelming users who only need to upload one document.

3. Accessibility of dynamic content
When files are added dynamically to the list, screen reader users need to be informed. An aria-live="polite" region that announces "File [name] added" and "File [name] removed" would make this pattern significantly more accessible. The current demos don't appear to address this.

4. Mobile considerations
Many government service users access services on mobile devices. The drag-and-drop zone is essentially unusable on mobile. The pattern should gracefully degrade to a tap-to-select experience without losing any functionality. I'd suggest the drag-and-drop zone be treated as an enhancement, not the primary interaction.

5. Upload size guidance
Users in government contexts often don't know what file sizes mean. Rather than showing "2MB" (which means nothing to most users), consider contextual guidance like "This file is too large. Try taking a photo of the document instead of scanning it."

Happy to share more detailed thinking on any of these points. Would be great to see this land as a pattern in the Design System.
